### PR TITLE
parse dimensions out of instance names

### PR DESCRIPTION
**Pull request by mdubbyap**

### CODE REVIEW

**Comment by cep21** [comment]
Please also update CHANGELOG.md

**Comment by mdubbyap** [comment]
@cep21  i believe i've addressed all your feedback.

File: protocol/collectd/collectd_test.go [hunk 1]
```diff
@@ -154,3 +154,24 @@ func TestCollectdJsonDecoding(t *testing.T) {
 	assert.Equal(t, "memory.free", dp.Metric)
 	assert.Equal(t, 3, len(dp.Dimensions))
 }
+
+func TestCollectdParseInstanceNameForDimensions(t *testing.T) {
+	type Case struct {
```
Comment (cep21) [inline protocol/collectd/collectd_test.go]: I like the use of anonymous structs for this.

File: protocol/collectd/collectd_test.go [hunk 2]
```diff
@@ -154,3 +154,24 @@ func TestCollectdJsonDecoding(t *testing.T) {
...
+		val string
+		dim map[string]string
+	}
+	tests := []Case{
+		{"name[k=v,f=x]", map[string]string{"k": "v", "f": "x", "instance": "name"}},
+		{"name[k=v,f=x]-rest", map[string]string{"k": "v", "f": "x", "instance": "name-rest"}},
+		{"name[k=v,f=x]-middle-[g=h]-more", map[string]string{"k": "v", "f": "x", "instance": "name-middle-[g=h]-more"}},
```
Comment (cep21) [inline protocol/collectd/collectd_test.go]: Can you add a test case for "name[k=v"

File: protocol/collectd/collectd.go [hunk 3]
```diff
@@ -84,6 +84,47 @@ func addIfNotNullOrEmpty(dimensions map[string]string, key string, cond bool, va
 	}
 }
 
+// try to pull out dimensions out of name in the format name[k=v,f=x]-morename would
+// return name-morename and extract dimensions (k,v) and (f,x)
+// if we encounter something we don't expect use original
+func getDimensionsFromName(val *string) (instanceName string, toAddDims map[string]string) {
```
Comment (cep21) [inline protocol/collectd/collectd.go]: You don't need to take val as a pointer.  Strings are themselves pointers to bytes similar to Java String.

Comment (cep21) [inline protocol/collectd/collectd.go]: You can still prefix the comment with "getDimensionsFromName ..." eve though it's private.

File: protocol/collectd/collectd.go [hunk 4]
```diff
@@ -63,10 +63,10 @@ func NewDatapoint(point *JSONWriteFormat, index uint, defaultDimensions map[stri
 	// Don't add empty dimensions
 	addIfNotNullOrEmpty(dimensions, "host", true, point.Host)
 	addIfNotNullOrEmpty(dimensions, "plugin", true, point.Plugin)
-	addIfNotNullOrEmpty(dimensions, "plugin_instance", true, point.PluginInstance)
+	parseInstanceNameForDimensions(dimensions, "plugin_instance", true, point.PluginInstance)
```
Comment (cep21) [inline protocol/collectd/collectd.go]: These can be nil.  I'll add the check.

**Pull request merged**